src/unix/aio.c

```diff
@@ -0,0 +1,352 @@
+#include "unix_internal.h"
+#include <page.h>
+
+#define AIO_RING_MAGIC  0xa10a10a1
+
+#define AIO_KNOWN_FLAGS IOCB_FLAG_RESFD
+
+#define AIO_RESFD_INVALID   -1U
+
+#define aio_lock(aio)   u64 _irqflags = spin_lock_irq(&(aio)->lock)
+#define aio_unlock(aio) spin_unlock_irq(&(aio)->lock, _irqflags)
+
+typedef struct aio_ring {
+    unsigned int id;
+    unsigned int nr;
+    unsigned int head;
+    unsigned int tail;
+    unsigned int magic;
+    unsigned int compat_features;
+    unsigned int incompat_features;
+    unsigned int header_length;
+    struct io_event events[0];
+} *aio_ring;
+
+struct aio {
+    struct list elem;  /* must be first */
+    heap h;
+    kernel_heaps kh;
```

Comment:
It's used not only to retrieve (via heap_general()) the heap for general allocations, but also to retrieve (via heap_pages() and heap_physical()) the heaps for process virtual memory mapping, so I figured I would store in the aio struct just a pointer to the kernel heaps and call the appropriate functions to retrieve the various heaps needed, instead of storing a pointer to each heap.